letsencrypt-aws.py

```diff
@@ -1,154 +1,266 @@
+                authz_record.change_id,
+                dns_challenge.validation_domain_name(authz_record.host),
+                dns_challenge.validation(acme_client.key),
+            )
+    private_key_bytes = private_key.private_bytes(
+        encoding=serialization.Encoding.PEM,
+        format=serialization.PrivateFormat.TraditionalOpenSSL,
+        encryption_algorithm=serialization.NoEncryption())
+    return Certificate(name,
+                       private_key_bytes,
+                       pem_certificate,
+                       pem_certificate_chain)
+
+
+def _upload_cert(cert, iam, path='/'):
+    return iam.create_server_certificate(
+        Path=path,
+        ServerCertificateName=cert.name,
+        CertificateBody=cert.cert_body,
+        PrivateKey=cert.key,
+        CertificateChain=cert.chain
+    )
 
-        # Sleep before trying to set the certificate, it appears to sometimes
-        # fail without this.
-        time.sleep(15)
-        logger.emit("updating-elb.set-elb-certificate", elb_name=self.elb_name)
-        self.elb_client.set_load_balancer_listener_ssl_certificate(
-            LoadBalancerName=self.elb_name,
-            SSLCertificateId=new_cert_arn,
-            LoadBalancerPort=self.elb_port,
-        )
+
+def _get_cert_arn(cert):
+    return cert.server_certificate_metadata['Arn']
+
+
+def _write_cert(cert):
+    with open(cert.name + '.key', 'w') as key:
+        key.write(cert.key)
+    with open(cert.name + '.crt', 'w') as crt:
+        crt.write(cert.cert_body)
+    with open(cert.name + '.chain', 'w') as chain:
+        chain.write(cert.chain)
+
+
+def _read_cert_from_disk(name):
+    with open(name + '.key') as key:
+        private_key = key.read()
+    with open(name + '.crt') as crt:
+        cert_body = crt.read()
+    with open(name + '.chain') as chain:
+        cert_chain = chain.read()
+    return Certificate(name, private_key, cert_body, cert_chain)
+
+
+def create(acme_client,
+           route53_client,
+           hosts,
+           path,
+           name,
+           save_local_copy=False):
+    iam = boto3.resource('iam')
+    challenge_completer = Route53ChallengeCompleter(route53_client)
+    logger.info('Creating certificate: {}'.format(name))
+    new_cert = _provision_cert(
+        name, hosts, acme_client, challenge_completer)
+    if(save_local_copy):
+        logger.info('Saving certificate {} to disk'.format(name))
+        _write_cert(new_cert)
+    logger.info('Uploading new {} certificate'.format(name))
+    _upload_cert(new_cert, iam, path)
+
+
+def update(acme_client,
+           route53_client,
+           path,
+           expiration_threshold,
+           regions,
+           read_local_copy=False,
+           save_local_copy=False):
+    iam = boto3.resource('iam')
+    challenge_completer = Route53ChallengeCompleter(route53_client)
+    for cert in _get_expiring_certificates(iam, path, expiration_threshold):
+        if read_local_copy:
```

Comment:
Might extract this code operating on a single cert into a separate method.  Would be much more easily tested if we ever want to write tests for this code.